sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java

```diff
@@ -219,6 +219,10 @@ public static void main(
     ShortIdMap metricsShortIds = new ShortIdMap();
     ExecutorService executorService =
         options.as(ExecutorOptions.class).getScheduledExecutorService();
+    // In order to reduce memory spent on per-thread coders and buffers, we separate finalizations
+    // from standard bundle processing.
+    ExecutorService finalizationExecutorService =
```

Comment:
This will be the same instance as the line above on 221

Reply:
Ah, I was confused by the default factory, but I guess that is just called once?
If I didn't use the option but just used UnboundedScheduledExecutorService directly, what would break? ie why is it an option?

If the thread pools are different, is there a benefit to having a single executor? We could share the task priority queue if we stored an additional executor with it.

Another idea was to try to periodically figure out how much memory the thread was pinning and drop it got large.
Even with separate groups, I'm concerned the per-thread objects could get big since they may just increase to the max observed.
Would figuring that out for a thread periodically that be possible with jamm? While we're hoping, if we could discover thread groups dynamically based upon call-site that would be cool but probably difficult to do efficiently.

Reply:
Yes the default factory is called once to populate the value the first time it is looked up if never set before.

One instance was used because it allowed users to configure it. Other then that there were some other minor benefits:
* it was used a lot for testing where we can inject things like the executor deep within the stack without needing to pass it through all the layers
* it simplified shutdown as we only had one executor to manage for the worker
* prevented people from incorrectly creating one since it isn't hard to incorrectly configure one where you get starved or spawn too many threads

Measuring memory thread pinning should be doable but enumerating thread locals is hacky: https://stackoverflow.com/questions/2001353/java-list-thread-locals

Using thread groups could be done in the same way where we are creating multiple executors but managed centrally via one so that each gets its own dedicated set of threads since you can create a named "sub-executor".

Can you share more about where we are using large thread locals? Maybe we can tackle that directly instead.

Reply:
The ones I noticed were related to AvroCoder.

There is a thread-local for the binarycoder here:
https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java#L299
and thread-local for GenericReader here:
https://github.com/rdblue/avro-java/blob/master/avro/src/main/java/org/apache/avro/generic/GenericDatumReader.java#L121

This was in Spanner->Avro GCS template pipeline. So we might be able to change the coder there to see gains. But it still seems like a generic issue with a single thread pool with distinct work types

Reply:
Avro and Beam should be using an object pool in both of these cases as the intent is to not pass information from one place to another in the same thread but to increase re-use.

We can see that Avro has a few memory leaks like https://issues.apache.org/jira/browse/AVRO-1283 and  https://issues.apache.org/jira/browse/AVRO-1595